### ISSUE TITLE
Cannot create or change Configurations Folder from FeatureIDE Project Properties

### ISSUE THREAD

**Issue by bjthehun** [FeatureIDE/FeatureIDE#1296]
### Prerequisites
- Eclipse Version: 2021-09 (4.21)
- FeatureIDE Version: 3.8.0
- Operating System: Ubuntu 21.04
- Java Runtime used to run Eclipse: OpenJDK 64-Bit Server VM 16.0.1
- FeatureIDE Composer: FeatureIDE Modelling

## Issue Description
In a Feature Modelling Project without a Configurations Folder, FeatureIDE throws the following exception when trying to declare an existing folder as the Configuration Folder:

```
java.lang.NullPointerException: Cannot invoke "org.eclipse.core.resources.IFolder.getProjectRelativePath()" because the return value of "de.ovgu.featureide.core.IFeatureProject.getConfigFolder()" is null
	at de.ovgu.featureide.core.propertypage.FeatureProjectPropertyPage.setPaths(FeatureProjectPropertyPage.java:406)
	at de.ovgu.featureide.core.propertypage.FeatureProjectPropertyPage.performOk(FeatureProjectPropertyPage.java:338)
	at org.eclipse.jface.preference.PreferencePage.performApply(PreferencePage.java:457)
	at org.eclipse.jface.preference.PreferencePage.lambda$1(PreferencePage.java:289)
	at org.eclipse.swt.events.SelectionListener$1.widgetSelected(SelectionListener.java:84)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5884)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1522)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5126)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4576)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.ui.dialogs.PropertyDialogAction.run(PropertyDialogAction.java:153)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:414)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5884)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1522)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5126)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4576)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1041)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
```

Instead, it creates the `build` folder, which is both unnecessary for a Feature Modeling Project, and cannot be removed from the Properties page.

Also, when a Configuration Folder already exists, FeatureIDE cannot change the folder, but fails with a different exception:

```
java.lang.IllegalArgumentException: Index out of bounds
	at org.eclipse.swt.SWT.error(SWT.java:4874)
	at org.eclipse.swt.SWT.error(SWT.java:4808)
	at org.eclipse.swt.SWT.error(SWT.java:4779)
	at org.eclipse.swt.widgets.Widget.error(Widget.java:560)
	at org.eclipse.swt.widgets.Combo.getItem(Combo.java:1166)
	at de.ovgu.featureide.core.propertypage.FeatureProjectPropertyPage.setCompositionMechanism(FeatureProjectPropertyPage.java:370)
	at de.ovgu.featureide.core.propertypage.FeatureProjectPropertyPage.performOk(FeatureProjectPropertyPage.java:341)
	at org.eclipse.jface.preference.PreferenceDialog$7.run(PreferenceDialog.java:905)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jface.util.SafeRunnable.run(SafeRunnable.java:174)
	at org.eclipse.jface.preference.PreferenceDialog.okPressed(PreferenceDialog.java:889)
	at org.eclipse.ui.internal.dialogs.FilteredPreferenceDialog.okPressed(FilteredPreferenceDialog.java:461)
	at org.eclipse.jface.preference.PreferenceDialog.buttonPressed(PreferenceDialog.java:233)
	at org.eclipse.jface.dialogs.Dialog.lambda$0(Dialog.java:619)
	at org.eclipse.swt.events.SelectionListener$1.widgetSelected(SelectionListener.java:84)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:252)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5884)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1522)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5126)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4576)
	at org.eclipse.jface.window.Window.runEventLoop(Window.java:823)
	at org.eclipse.jface.window.Window.open(Window.java:799)
	at org.eclipse.ui.dialogs.PropertyDialogAction.run(PropertyDialogAction.java:153)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:474)
	at org.eclipse.jface.action.ActionContributionItem.handleWidgetSelection(ActionContributionItem.java:580)
	at org.eclipse.jface.action.ActionContributionItem.lambda$4(ActionContributionItem.java:414)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5884)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1522)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:5126)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4576)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1041)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1440)
```

### Steps to Reproduce

1. In an existing Feature Modelling project (here named `BerkeleyDB`), create a new folder named `configs`. The project structure now is:
![ProjectViewBefore](https://user-images.githubusercontent.com/38729215/138075240-35dcc5dd-c3ea-41ac-90e1-f01bceaac3d6.png)

2. Right click on the project in the project explorer, then click `Properties` > `FeatureIDE` > `Feature Project`. The text fields `Feature path` and `Source path` are greyed out:
![SetConfigFolder](https://user-images.githubusercontent.com/38729215/138075250-e19a4ea0-1be0-4aa6-b462-9f6e66fff9fb.png)

3. Type `configs` under the `configurations path` text field, and then press `Apply`. This results in the first exception, and also creates the `build` folder.

4. Create a new configuration named `DBConfig.xml` config. FeatureIDE creates a new folder named `equations`, declares it as configuration folder, and stores `DBConfig.xml` there:
![ActualProjectStructure](https://user-images.githubusercontent.com/38729215/138075286-febd56cd-949f-45fc-bb0a-60f5cfaebb2a.png)

 
5. Attempt to change the configuration folder to `configs` as described in Steps 2 and 3, resulting in the second exception.

**Comment by jabiercoding** [comment]
This bug makes that I cannot change anything in the project preferences (of a Munge project in case it is important).
I think this bug should be considered "major" instead of minor.

The fix can be: In FeatureProjectPropertyPage
https://github.com/FeatureIDE/FeatureIDE/blob/05469445f364f09474a98a0bd438f8b0f71e7218/plugins/de.ovgu.featureide.core/src/de/ovgu/featureide/core/propertypage/FeatureProjectPropertyPage.java#L370

change
`		final String item = mechanismCombo.getItem(mechanismCombo.getSelectionIndex());`
with something like
```		
		final int index = mechanismCombo.getSelectionIndex();
		if (index == -1) {
			return;
		}
		final String item = mechanismCombo.getItem(index);
```

**Comment by skrieter** [comment]
I made a fix in 09623f0e16fbed2cfcd9c3e238ec397d737a0625 by checking and handling null reference at corresponding locations in the code. However, as a more reliable solution, we should consider to change the return values of the properties of the feature project (e.g., config folder location, source folder location, etc.) to `Optional<>`. This way, developers are forced to account for missing values, for instance when dealing with feature modeling projects.
